File: 3076-shortest-uncommon-substring-in-an-array/3076-shortest-uncommon-substring-in-an-array.cpp
class Solution {
public:
    vector<string> shortestSubstrings(vector<string>& arr) {
        
        int n = arr.size();
        
        unordered_map<string,vector<string>> mp1;
        unordered_map<string,int> mp2;
        
        for(int i=0; i<n; i++) {
            unordered_set<string> st;
            for(int x=0; x<arr[i].size(); x++) {
                string str="";
                for(int y=x; y<arr[i].size(); y++) {
                    str += arr[i][y];
                    mp1[arr[i]].push_back(str);
                    st.insert(str);
                }
            }
            for(string it: st)  mp2[it]++;
        }
        
        vector<string> ans(n);
        
        for(int i=0; i<n; i++) {
            string str="";
            for(auto it: mp1[arr[i]]) {
                if(mp2[it] == 1) {
                    if(str=="") str = it;
                    else {
                        if(str.length() == it.length() && str > it) {
                            str = it;
                        }
                        else if(str.length() > it.length()) {
                            str = it;
                        }
                    }
                    
                }
            }
            ans[i] = str;
        }
        
        return ans;
    }
};

File: 3076-shortest-uncommon-substring-in-an-array/3076-shortest-uncommon-substring-in-an-array.py
class Solution:
    def shortestSubstrings(self, arr: List[str]) -> List[str]:
        if not arr:
            return arr
        hashmap = dict()
        for i, word in enumerate(arr):
            for size in range(1, len(word)+1):
                for sli in range(len(word) - size + 1):
                    temp = word[sli:sli+size]
                    if i not in hashmap:
                        hashmap[i] = set()
                    hashmap[i].add(temp)
        rst = {k: list(v) for k, v in hashmap.items()}
        for i, d in enumerate(rst):
            for i_com in range(len(hashmap)):
                if i_com == i:
                    continue
                for v_com in hashmap[i_com]:
                    if v_com in rst[i]:
                        rst[i].remove(v_com)
        rst = {k: [(len(s), s) for s in v] for k, v in rst.items()}
        rst = [sorted(s)[0][1] if len(s) > 0 else '' for s in rst.values()]
        return rst